pipeline_dp/contribution_bounders.py

```diff
@@ -141,3 +142,73 @@ def rekey_per_privacy_id_per_partition_key(pid_pk_dict):
         return backend.map_values(
             col, aggregate_fn,
             "Apply aggregate_fn after per privacy id contribution bounding")
+
+
+class SamplingCrossPartitionContributionBounder(ContributionBounder):
+    """Bounds the contribution of privacy_id cross partitions.
+
+    It ensures that each privacy_id contributes to not more than
+    max_partitions_contributed partitions (cross-partition contribution
+    bounding), by performing sampling if needed. It is assumed that provided
+    aggregate_fn function performs per-partition contributions.
+    """
+
+    def bound_contributions(self, col, params, backend, report_generator,
+                            aggregate_fn):
+        col = backend.map_tuple(
+            col, lambda pid, pk, v: (pid, (pk, v)),
+            "Rekey to ((privacy_id), (partition_key, value))")
+
+        col = collect_values_per_partition_key_per_privacy_id(col, backend)
+        # (privacy_id, [partition_key, [value]])
+
+        # Bound cross partition contributions with sampling.
+        sample = sampling_utils.choose_from_list_without_replacement
+        sample_size = params.max_partitions_contributed
+        col = backend.map_values(col, lambda a: sample(a, sample_size))
+
+        # (privacy_id, [partition_key, [value]])
+
+        # Unnest the list per privacy id.
+        def rekey_per_privacy_id_per_partition_key(pid_pk_v_values):
+            privacy_id, partition_values = pid_pk_v_values
+
+            for partition_key, values in partition_values:
+                yield (privacy_id, partition_key), values
+
+        col = backend.flat_map(col, rekey_per_privacy_id_per_partition_key,
+                               "Unnest per privacy id")
+        # ((privacy_id, partition_key), [value])
+
+        return backend.map_values(
+            col, aggregate_fn,
+            "Apply aggregate_fn after per privacy id contribution bounding")
+
+
+def collect_values_per_partition_key_per_privacy_id(
```

Comment:
There seems to be a naming collision with an inline function for SamplingPerPrivacyId, which is not great also because that means we are not able to effectively explain the difference between these two very similar functions. I think we should

1. Name them differently
2. Consider if we can refactor to potentially reuse some code? Maybe SamplingPerPrivacyId could be made to work with this new function. Going even further, we have [similar](https://github.com/google/differential-privacy/blob/main/privacy-on-beam/pbeam/aggregations.go#L870)* helper functions in Privacy on Beam that are shared throughout the codebase. Maybe there is an opportunity to do this for python as well?

* Functions that convert different permutations of (privacy_id, partition_key, value) to other permutation of (privacy_id, partition_key, value), expanding values to lists or unnesting values...

Reply:
Thanks, that's very good point. I've changed `collect_values_per_partition_key_per_privacy_id` to be used in `SamplingPerPrivacyId`. Also I made usage of this function in utility analysis contribution bounder as well. PTAL

Reply:
Looks great! One minor comment: We are now calling the pipeline_dp.contribution_bounders from utility_analysis_new.contribution_bounders. Do you think there is another place more fitting for such a shared function?

Reply:
SG, thanks!